README.md
# Terminal Macros

*Use the terminal without using the terminal*

## Add Your Commands

Add and organize all of your commands.

 - settings.json

```jsonc
"terminalMacros.commands": [
	{
		"command": "echo Hello world"
	},
	{
		// Required
		"command": "{recent}", // {recent} = the command that was run most recently

		// Optional
		"group": "General", // Used for organization
		"name": "Last Command", // The pretty name
		"save": true, // Saves your file
		"show": true, // Shows the terminal
		"stop": false, // Control c
		"logout": false, // Control d
		"clear": true, // Clear the terminal
		"execute": true, // Execute the command
		"focus": false, // Focus on the terminal
		"terminal": "bash" // Name of terminal to use
	},
	{
		"group": "General",
		"name": "Run File",
		"command": "./{file}" // {file} = name of the current file in the editor
	},
	{
		"group": "General",
		"name": "Paste",
		"command": "{paste}" // {paste} = the last thing copied
	},
	{
		"group": "General",
		"name": "Full Path",
		"command": "echo {path}" // {path} = full path of the current file in the editor
	},
	{
		"group": "General",
		"name": "Execute Selected",
		"command": "{selection}", // {selection} = selected text in the editor
		"execute": true
	},
	{
		"group": "General",
		"name": "Change Directory",
		"command": "cd {directory}", // {directory} = full path of the directory of the file in the editor
		"execute": true
	}
],
```

## Edit Default Command Options

Useful if you always want to clear the terminal, save your file, etc.

 - Settings

 - settings.json

```jsonc
"terminalMacros.default.clear": true,
"terminalMacros.default.execute": true,
"terminalMacros.default.focus": false,
"terminalMacros.default.logout": false,
"terminalMacros.default.save": true,
"terminalMacros.default.show": true,
"terminalMacros.default.stop": false,
```

## Execute a Command with Keyboard Shortcuts

Use your most frequent commands with keyboard shortcuts.

 - keybindings.json

```jsonc
{
	"key": "f1",
	"command": "terminalMacros.executeCommand",
	"args": 0 // Required: command number, a 0-based index in the array
},
```

## List All Commands in Quick Pick Menu

See all your commands in the quick pick menu.

 - Command Palette: "Terminal Macros: List Commands"

 - Keyboard Shortcuts

 - keybindings.json

```jsonc
{
	"key": "f2",
	"command": "terminalMacros.listCommands"
},
```


CHANGELOG.md
# Change Log

## [1.0.15] - 2023 Oct 25
- Fix: Remove unnecessary files

## [1.0.14] - 2023 Mar 1
- Fix: Terminals created without a shell executable (Remote Explorer)

## [1.0.13] - 2023 Feb 7
- Fix: Improve cmd, powershell, and wsl terminals changing names

## [1.0.12] - 2023 Feb 3
- Fix: Clearing terminals with scrollbars

## [1.0.11] - 2023 Jan 27
- Performance: Sped up executeCommand and listCommands
- Info: Simplified error messages

## [1.0.10] - 2023 Jan 26
- Info: Fixed descriptions of stop & logout in settings

## [1.0.9] - 2022 Dec 14
- Add: Variable {directory} in command string to get the full path of the current directory
- Fix: Issue with paths for WSL

## [1.0.8] - 2022 Dec 2
- Info: Changed icon
- Info: Fixed typos in README.md

## [1.0.7] - 2022 Dec 2
- Fix: Issue with options per command

## [1.0.6] - 2022 Dec 1
- Add: Variable {path} to command string to get the full path of the current file
- Add: Variable {selection} to command string to get the selected text
- Add: Option "terminal" to individual/default commands to choose terminal name
- Fix: Issue with command being cut off with WSL, stop/logout, and clear

## [1.0.5] - 2022 Feb 8
- Info: Remove known issue because VSCode fixed inability to open terminal without focus

## [1.0.4] - 2021 Sep 29
- Fix: issue with ms-vscode-remote.remote-ssh

## [1.0.3] - 2021 Sep 29
- Fix: settings.json bug with terminalMacros.commands

## [1.0.2] - 2021 Sep 19
- Info: Updated description

## [1.0.1] - 2021 Sep 19
- Info: Fixed wording and typos in README.md

## [1.0.0] - 2021 Sep 19
- Initial release


src/menu.js
import * as vscode from 'vscode';

interface TerminalCommand {
	command: string;
	auto: boolean;
	preserve: boolean;
	name?: string;
	group?: string;
}

export interface CommandQuickPickItem extends vscode.QuickPickItem {
	type: 'group' | 'command';
	command?: TerminalCommand;
	group?: string;
}

export async function showCommandsPick(commands) {
	const grouplessCommands = getGroupless(commands);
	const groups = getGroups(commands);

	const pickItems = getPickItems(grouplessCommands, groups);
	if (pickItems.length === 0) {
		return;
	}

	let picked = await vscode.window.showQuickPick(pickItems, { matchOnDescription: true });
	if (!picked) {
		return;
	}

	if (picked.type === 'group' && picked.group) {
		picked = await vscode.window.showQuickPick(
			getPickItems(getFromGroup(commands, picked.group)),
			{ matchOnDescription: true });
	}

	if (!picked || picked.type !== 'command' || !picked.command) {
		return;
	}

	return picked.command;
}

function getGroupless(commands) {
	return commands.filter(c => !c.group);
}

function getGroups(commands) {
	return distinct(commands
		.filter(c => !!c.group)
		.map(c => c.group as string));
}

function getFromGroup(commands, group) {
	return commands.filter(c => c.group === group);
}

function distinct<T>(values) {
	return [...new Set(values)];
}

function getPickItems(commands, groups = []) {
	return [
		...commands.map<CommandQuickPickItem>(c => {
			return {
				type: 'command',
				command: c,
				label: c.name || removeVariables(c.command).trim(),
				description: 'Command' + (c.name ? ` (${removeVariables(c.command).trim()})` : '')
			};
		}),
		...groups.map<CommandQuickPickItem>(g => {
			return {
				type: 'group',
				group: g,
				label: g,
				description: 'Group'
			};
		})
	];
}

function removeVariables(command) {
	return command.replace(/{\w+}/g, '');
}

extension.js
const vscode = require('vscode');

const path = require('node:path');

const CODES = {
	escape:        '\u001b',
	up:            '\u001b[A',
	stopProcess:   '\03', // control c
	logout:        '\04', // control d
	clearBefore:   '\13',
	clearTerminal: '\14',
	paste:         '\26',
	clearAfter:    '\27',
};

const TERMINALS = {
	cmd:        1,
	powershell: 2,
	wsl:        3,
};

//
// Helper Functions
//

function error(message) {
	vscode.window.showErrorMessage(`Terminal Macros: ${message}`);
}

function getConfig() {
	return vscode.workspace.getConfiguration().get('terminalMacros');
}

function windowsToWslPath(fileName) {
	return fileName.split(path.sep).join(path.posix.sep).replace(/^([a-zA-Z]):/, '/mnt/$1');
}

async function formatCommand(commandText, clear, execute, specialTerminal) {
	// Replace {recent} format specifier and do nothing else
	if (/{recent}/.test(commandText)) {
		commandText = CODES.up;

		// Fix for clearing the screen and then getting a recent command in cmd
		if (specialTerminal == TERMINALS.cmd && clear && execute)
			commandText += CODES.up;

		return commandText;
	}

	// Replace {paste} format specifier using the clipboard
	commandText = await commandText.replaceAll('{paste}', await vscode.env.clipboard.readText());

	// Replace format specifiers for the editor
	const editor = vscode.window.activeTextEditor;
	if (editor) {
		const document = editor.document;
		let fileName = document.fileName;

		// Fix for WSL path
		if (specialTerminal == TERMINALS.wsl)
			fileName = windowsToWslPath(fileName);

		commandText = commandText.replaceAll('{directory}', path.dirname(fileName));
		commandText = commandText.replaceAll('{file}', path.basename(fileName));
		commandText = commandText.replaceAll('{path}', fileName);
		commandText = commandText.replaceAll('{selection}', document.getText(editor.selection));
	}

	return commandText;
}

async function prepareTerminal(terminal, stop, logout, clear, specialTerminal) {
	// Clear line
	if (specialTerminal == TERMINALS.cmd || specialTerminal == TERMINALS.powershell)
		terminal.sendText(CODES.escape, false);
	else
		terminal.sendText(CODES.clearBefore + CODES.clearAfter, false);

	// Stop process
	if (stop)
		terminal.sendText(CODES.stopProcess, false);

	// Log out of session
	if (logout)
		terminal.sendText(CODES.logout, false);

	// Clear terminal
	if (clear) {
		// Clear previous content on screen
		await vscode.commands.executeCommand('workbench.action.terminal.clear');

		// Clear control codes on screen
		if (specialTerminal == TERMINALS.cmd)
			terminal.sendText('cls');
		else
			terminal.sendText(CODES.clearTerminal, false);
	}
}

function getSpecialTerminalEnum(terminal) {
	// If terminal not created by a path in this filesystem, then skip
	const shellPath = terminal.creationOptions.shellPath;
	if (!shellPath)
		return;

	// If not a Windows executable, then skip
	if (path.extname(shellPath) != '.exe')
		return;

	// Get the name of the Windows executable
	const name = path.parse(shellPath).name;
	return TERMINALS[name];
}

//
// Extension Commands
//

async function executeCommand(n, config=undefined) {
	// Get all commands
	if (config == undefined)
		config = getConfig();
	const commands = config.commands;
	const len = commands.length;

	// Error if no command
	if (len == 0)
		return error('settings.json: No commands');
	else if (isNaN(n) || n < 0 || n >= len)
		return error(`keybindings.json: "args": ${n} not command number in range 0 - ${len-1} inclusive`);

	// Get command
	const command = commands[n];
	const group = command.group;
	const name = command.name;
	let commandText = command.command;

	// Error in no command text
	if (commandText == undefined)
		return error('settings.json: "command": "COMMAND_HERE" missing');

	// Get command options
	const defaults = config.default;
	const save = (command.save != undefined) ? command.save : defaults.save;
	const show = (command.show != undefined) ? command.show : defaults.show;
	const stop = (command.stop != undefined) ? command.stop : defaults.stop;
	const logout = (command.logout != undefined) ? command.logout : defaults.logout;
	const clear = (command.clear != undefined) ? command.clear : defaults.clear;
	const execute = (command.execute != undefined) ? command.execute : defaults.execute;
	const focus = (command.focus != undefined) ? command.focus : defaults.focus;
	const terminalName = command.terminal || defaults.terminal;

	// Get terminal
	let terminal;
	// Get specific terminal
	if (terminalName) {
		// Find it
		for (existingTerminal of vscode.window.terminals) {
			if (existingTerminal.name == terminalName) {
				terminal = existingTerminal;
				break;
			}
		}

		// Create it
		if (!terminal)
			terminal = vscode.window.createTerminal({ name: terminalName });
	}
	// Get any terminal
	else {
		// Find it
		terminal = vscode.window.activeTerminal;

		// Create it
		if (!terminal)
			terminal = vscode.window.createTerminal();
	}

	// Show terminal
	if (show)
		terminal.show(!focus);

	// Save file
	if (save && vscode.window.activeTextEditor)
		await vscode.window.activeTextEditor.document.save();

	const specialTerminal = getSpecialTerminalEnum(terminal);

	// Prepare terminal before executing command
	await prepareTerminal(terminal, stop, logout, clear, specialTerminal);

	// Prepare command
	commandText = await formatCommand(commandText, clear, execute, specialTerminal);

	// Fix for powershell consuming input while stopping
	const wait = specialTerminal == TERMINALS.powershell && stop;

	// Send command
	setTimeout(function() {
		terminal.sendText(commandText, execute);
	}, wait ? 5 : 0);
}

async function listCommands(currentGroup = undefined) {
	// Get all commands
	const config = getConfig();
	const commands = config.commands;

	// Make a list of QuickPickItem objects
	const groupNames = {};
	const quickPickItems = [];
	for (let n = 0; n < commands.length; n++) {
		const command = commands[n];

		// Add commands
		if (command.group == currentGroup) {
			quickPickItems.push({
				// Command data
				'isCommand': true,
				'n': n,

				// QuickPickItem data
				'label': command.name || command.command,
				'description': (command.name == undefined) ? '': command.command
			});
		}
		// Add groups
		else if (currentGroup == undefined && ! (command.group in groupNames)) {
			quickPickItems.push({
				// Group data
				'isCommand': false,

				// QuickPickItem data
				'label': command.group
			});
			groupNames[command.group] = true;
		}
	}

	// Wait for the user
	const picked = await vscode.window.showQuickPick(quickPickItems);
	if (!picked)
		return;

	// Execute the command
	if (picked.isCommand)
		executeCommand(picked.n, config);
	// Recursively display the groups
	else
		listCommands(picked.label);
}

//
// Basic Extension Functions
//

function activate(context) {
	// Implement activationEvents here, defined in package.json
	// activationEvents can have keyboard shortcuts, and commands show up in the command palette
	context.subscriptions.push(
		vscode.commands.registerCommand('terminalMacros.executeCommand', executeCommand),
		vscode.commands.registerCommand('terminalMacros.listCommands', listCommands),
	);
}

function deactivate() {}

module.exports = {
	activate,
	deactivate
}
